### ISSUE TITLE
Extend admin server library: support callback that are invoked when command complete

### ISSUE THREAD

**Issue by per-samuelsson** [Starcounter/Starcounter#2587]
In #2568, the admin server library was extended to support an optional callback to be given when executing state changing command to be cancelled.

Another requirement is to support a callback mechanism that can be used to detect when a command has completed (successfully, cancelled, or failed).

**Comment by per-samuelsson** [comment]
The API will be like this:

```C#
var server = GetServerReference();  // IServerRuntime

var command = new StartDatabaseCommand("Foo");
var started = DateTime.Now;
var result = server.Execute(command, null, (c) => {
  var duration = DateTime.Now - started;
  Console.WriteLine("Command {0} took {1} to complete", c, duration);
});
```

**Comment by per-samuelsson** [comment]
Developed and merged to develop in a set of commits merged in https://github.com/Starcounter/Starcounter/commit/b535671909a8635fa032e2d386ff31e3ef425d34.

**Comment by andwah** [comment]
@per-samuelsson When i get a callback, how do i check if the command has any errors?
The callback only give me access to the ServerCommand not the CommandInfo instance.

**Comment by per-samuelsson** [comment]
> The callback only give me access to the ServerCommand not the CommandInfo instance

Good point. If you need access to the command information, you need to get the initial one back from Execute. And you haven't got that when you pass in the delegate.

I need to consider this some. Maybe we need a redesign.

**Comment by per-samuelsson** [comment]
> I need to consider this some. Maybe we need a redesign.

One idea is that instead of passing the ```ServerCommand```, I'll pass the final ```CommandInfo``` instead. And maybe change that in the cancellation predicate too?

Another idea is to change so that a completion delegate is passed to an overloaded ```Wait``` method, not to the ```Execute``` method, like

```C#
var server = GetServerReference();  // IServerRuntime

var command = new StartDatabaseCommand("Foo");
var handle = server.Execute(command);
server.Wait(handle, (commandInfo) => {
  Console.Write("Im done");
});
```

Not sure what approach is best actually.

**Comment by per-samuelsson** [comment]
> Not sure what approach is best actually.

I will change so that I pass the unique command ID to both callbacks instead. With the ID, you can easily get the command info if you need it, and if you don't, the performance penalty is zero.

**Comment by per-samuelsson** [comment]
> I will change so that I pass the unique command ID to both callbacks instead. With the ID, you can easily get the command info if you need it, and if you don't, the performance penalty is zero.

This is now done, in #2589. The new callbacks are based on unique command ID's. In your case, you can get the final state of the command with code like this

```C#
var command = new StartDatabaseCommand("Foo");
var commandInfo = server.Execute(command, null, (id) => {
  var result = server.GetCommand(id);
  if (result.HasErrors) { ... }
});
```

**Comment by andwah** [comment]
@per-samuelsson Thanks, great news.

**Comment by andwah** [comment]
@per-samuelsson btw, is there a way to know what the database is working on?.  i need to know it the database is "starting" or "stopping"? 
(So i dont execute a StopDatabaseCommand on a stopping database etc..)

**Comment by andwah** [comment]
If a command (in my case the StarDatabaseCommand) is diectly cancelled, then the completed callback is never called.

**Comment by per-samuelsson** [comment]
> If a command (in my case the StarDatabaseCommand) is diectly cancelled, then the completed callback is never called.

Confirmed. This is a bug on my part. I'll fix it within a day. Thanks for reporting.

**Comment by per-samuelsson** [comment]
> Confirmed. This is a bug on my part. I'll fix it within a day. Thanks for reporting.

Working to fix this now, seperate issue #2601.
